hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs

```diff
@@ -117,68 +118,37 @@ loadCradle iniDynFlags (NewCradle fp) = do
   traceShowM cradle
   liftIO (GHC.newHscEnv iniDynFlags) >>= GHC.setSession
   liftIO $ setCurrentDirectory (BIOS.cradleRootDir cradle)
-  res <- gcatches
-    (do
-      withProgress "Initialising Cradle" NotCancellable (initializeCradle cradle)
-      return $ IdeResultOk ()
-    )
-    [ ErrorHandler $
-      \(err :: GHC.GhcException) -> do
+  res <- withProgress "Initialising Cradle" NotCancellable (\f ->
+          BIOS.initializeFlagsWithCradleWithMessage (Just (toMessager f)) fp (fixCradle cradle)
+         )
+  case res of
+    BIOS.CradleNone -> return (IdeResultOk ())
+    BIOS.CradleFail err -> do
       logm $ "GhcException on cradle initialisation: " ++ show err
       return $ IdeResultFail $ IdeError
           { ideCode    = OtherError
           , ideMessage = Text.pack $ show err
           , ideInfo    = Aeson.Null
           }
-    , ErrorHandler $  \(err :: IOException) -> do
-      logm $ "IOException on cradle initialisation: " ++ show err
-      return $ IdeResultFail $ IdeError
-          { ideCode    = OtherError
-          , ideMessage = Text.pack $ show err
-          , ideInfo    = Aeson.Null
-          }
-    , ErrorHandler $  \(err :: ErrorCall) -> do
-      logm $ "ErrorCall on cradle initialisation: " ++ show err
-      return $ IdeResultFail $ IdeError
-          { ideCode    = OtherError
-          , ideMessage = Text.pack $ show err
-          , ideInfo    = Aeson.Null
-          }
-    ]
-
-
-  case res of
-    IdeResultOk () -> do
+    BIOS.CradleSuccess init -> do
```

Comment:
Does this catch exceptions as well? E.g., if the cabal file is not correct, or the project fails to compile?

Reply:
I feel like we should catch these exceptions earlier than this.

Which bit throws an error if the cabal file isn't correct? `hie-bios` should no longer do that.

Having checked and unchecked exceptions isn't good.

Reply:
Should we catch the exceptions here now?